architecture-decisions/2024-06-06_arc4-by-default.md

```diff
@@ -0,0 +1,153 @@
+   }
+}
+```
+
+Handling of 'other' on completion actions can be done via decorators however the prescribed way in the documentation is to implement a [well-known](https://tealscript.netlify.app/guides/lifecycle/) method on your contract.
+
+Pros:
+ - A developer new to Algorand does not need to learn much about ARC4 in order to implement an ARC4 compliant contract.
+ - Flexibility exists for advanced devs to 'break out' of the ARC4 default when desired (kind of)
+
+Cons:
+ - The current implementation in TealScript outputs a lot of redundant teal in scenarios where an advanced user wants to do their own routing
+ - Doesn't give us a low friction path forward if ARC4 were to be replaced, or significantly updated
+ - In TypeScript, class members are public by default which could lead to accidentally exposing a method which shouldn't be.
+
+
+### Option 2 - Minimum Viable Smart Contract by default, ARC4 is opt in
+
+The approach taken by Alogrand Python is that a smart contract which extends the base `algopy.Contract` class will inherit no behaviour beyond that described in the context above (ie. `onComplete == clearState ? clearStateProgram() : approvalProgram()`).  This means the minimum viable smart contract (MVSC) would look something like this:
+
+```python
+# Contract which always approves
+class MVContract(Contract):
+    def approval_program(self) -> bool:
+        return True
+
+    def clear_state_program(self) -> bool:
+        return True
+```
+
+a TypeScript version might look like this:
+
+```ts
+export default class HelloWorldContract extends Contract {
+   approvalProgram(): boolean {
+      return true
+   }
+
+   clearState(): boolean {
+      return true
+   }
+}
+
+```
+
+An ARC4 compliant contract would instead extend the `algopy.arc4.ARC4Contract` base class. Each method must be decorated with either `@abimethod` or `@subroutine` to indicate a public or private method. 
+
+```python
+class HelloWorldContract(ARC4Contract):
+    @arc4.abimethod
+    def hello(self, name: String) -> String:
+        return "Hello, " + name
+```
+
+or in TypeScript
+
+```ts
+export default class HelloWorldContract extends arc4.Arc4Contract {
+  @arc4.abimethod()
+  sayHello(name: str): arc4.Str {
+    return new arc4.Str(`Hello ${name}`)
+  }
+}
+```
+
+Pros:
+- ARC4 support is built on top of a framework that can function in its own right giving us greater confidence we can support the next ARC if/when it comes, without breaking existing ARC4 support
+- MVSC gives power users absolute control without ARC4 getting in the way
+- Explicit decorators reduce the likelihood of accidentally exposing a method publicly.
+
+Cons:
+- A developer needs to consciously choose ARC4 which means they need to be aware of it beforehand, and require a basic understanding of its workings
+- Requiring decorators on all methods adds noise to what might otherwise be a very terse contract.
+
+### Option 3 - Option 2, but use public/private keywords
```

Comment:
This entire option is a bit confusing to me. If anything it seems more like option 1 because it's implicitly ABI without a decorator. It also would require every contract to have at least one ARC4 method as an entry point even if the developer didn't want actually use ARC4.

Reply:
The difference is the base contract. TealScript only has one base contract which opts you into arc4 by default. Option 2 and 3 have `Contract` and `ARC4Contract` base classes. If a user has extended the `ARC4Contract` class then it's safer to assume they want public methods to be abi routed. I think this option gives us the best of both worlds as users can opt out of arc4 entirely by extending `Contract` and opt in with `ARC4Contract`

Reply:
Ah ok yeah sorry I missed that. I like this option then and it's basically how TEALScript works. Public methods are ABI methods and private/protected are subroutines.